modules/boulder/src/main/ts/ephox/boulder/core/ValueProcessor.ts

```diff
@@ -1,262 +1,209 @@
-  return field.fold(
-    (key, okey, presence, prop) => {
-      const bundle = (av: any): SimpleResult<SchemaError[], any> => {
-        const result = prop.extract(path.concat([ key ]), strength, av);
-        return SimpleResult.map(result, (res) => ObjWriter.wrap(okey, strength(res)));
+type SimpleBundle = SimpleResult<SchemaError[], any>;
+type OptionBundle = SimpleResult<SchemaError[], Record<string, Optional<any>>>;
+
+const cExtractOne = <T>(path: string[], obj: Record<string, T>, value: ValuePresence.ValueProcessorTypes): SimpleResult<SchemaError[], T> => {
+  return ValuePresence.fold(
+    value,
+    (key, newKey, presence, prop) => {
+      const bundle = (av: any): SimpleBundle => {
+        const result = prop.extractProp(path.concat([ key ]), av);
+        return SimpleResult.map(result, (res) => ObjWriter.wrap(newKey, res));
       };
 
-      const bundleAsOption = (optValue: Optional<any>): SimpleResult<SchemaError[], Record<string, Optional<any>>> => {
+      const bundleAsOption = (optValue: Optional<any>): OptionBundle => {
         return optValue.fold(() => {
-          const outcome = ObjWriter.wrap(okey, strength(Optional.none()));
+          const outcome = ObjWriter.wrap(newKey, Optional.none());
           return SimpleResult.svalue(outcome);
         }, (ov) => {
-          const result: SimpleResult<any, any> = prop.extract(path.concat([ key ]), strength, ov);
+          const result: SimpleResult<any, any> = prop.extractProp(path.concat([ key ]), ov);
           return SimpleResult.map(result, (res) => {
-            return ObjWriter.wrap(okey, strength(Optional.some(res)));
+            return ObjWriter.wrap(newKey, Optional.some(res));
           });
         });
       };
 
-      return (() => {
-        return presence.fold(() => {
-          return SimpleResult.bind(
-            strictAccess(path, obj, key),
-            bundle
-          );
-        }, (fallbackThunk) => {
-          return SimpleResult.bind(
-            fallbackAccess(obj, key, fallbackThunk),
-            bundle
-          );
-        }, () => {
-          return SimpleResult.bind(
-            optionAccess(obj, key),
-            bundleAsOption
-          );
-        }, (fallbackThunk) => {
-          // Defaulted option access
-          return SimpleResult.bind(
-            optionDefaultedAccess(obj, key, fallbackThunk),
-            bundleAsOption
-          );
-        }, (baseThunk) => {
-          const base = baseThunk(obj);
-          const result = SimpleResult.map(
-            fallbackAccess(obj, key, Fun.constant({})),
-            (v) => Merger.deepMerge(base, v)
-          );
-          return SimpleResult.bind(result, bundle);
-        });
-      })();
+      const processPresence = (presence: FieldPresence.FieldPresenceTypes) => {
+        switch (presence.discriminator) {
+          case 'strict':
```

Comment:
Maybe use enums here as well. I like enums over strings since they have more context they can be renamed using ide tools they clump together as a type they compile to numbers etc.

Reply:
- [ ] do enums?